api/src/paths/project/{projectId}/survey/{surveyId}/observations/upload.ts

```diff
@@ -0,0 +1,326 @@
+                                oneOf: [
+                                  {
+                                    type: 'string'
+                                  },
+                                  {
+                                    type: 'number'
+                                  }
+                                ]
+                              }
+                            }
+                          }
+                        },
+                        rowErrors: {
+                          type: 'array',
+                          items: {
+                            type: 'object',
+                            properties: {
+                              type: {
+                                type: 'string'
+                              },
+                              code: {
+                                type: 'string'
+                              },
+                              message: {
+                                type: 'string'
+                              },
+                              row: {
+                                type: 'number'
+                              }
+                            }
+                          }
+                        },
+                        isValid: {
+                          type: 'boolean'
+                        }
+                      }
+                    }
+                  }
+                ]
+              }
+            }
+          }
+        }
+      }
+    },
+    400: {
+      $ref: '#/components/responses/400'
+    },
+    401: {
+      $ref: '#/components/responses/401'
+    },
+    403: {
+      $ref: '#/components/responses/401'
+    },
+    500: {
+      $ref: '#/components/responses/500'
+    },
+    default: {
+      $ref: '#/components/responses/default'
+    }
+  }
+};
+
+function prepDWCArchive(): RequestHandler {
+  return async (req, res, next) => {
+    defaultLog.debug({ label: 'prepDWCArchive', message: 'files.length', files: req?.files?.length });
+
+    if (!req.params.surveyId) {
+      throw new HTTP400('Missing required path param: surveyId');
+    }
+
+    if (!req.files || !req.files.length) {
+      // no media objects included
+      throw new HTTP400('Missing upload data');
+    }
+
+    if (req.files.length !== 1) {
+      // no media objects included
+      throw new HTTP400('Too many files uploaded, expected 1.');
+    }
+
+    try {
+      const rawMediaArray: Express.Multer.File[] = req.files as Express.Multer.File[];
+
+      const rawMediaFile = rawMediaArray[0];
+
+      const mediaFiles = parseUnknownMedia(rawMediaFile);
+
+      const dwcArchive = new DWCArchive(mediaFiles);
+
+      req['dwcArchive'] = dwcArchive;
+
+      next();
+    } catch (error) {
+      defaultLog.debug({ label: 'prepDWCArchive', message: 'error', error });
+      throw error;
+    }
+  };
+}
+
+/**
+ * Validate a Darwin Core csv file.
+ *
+ * @returns {RequestHandler}
+ */
+function validateDWCArchive(): RequestHandler {
```

Comment:
question:

the prep function and this function in this file seem very similar to the one inside `validate.ts`
can they not be reused?

Reply:
Its pretty similar yeah, technically the one in upload also checks for a survey Id, which is different.  There is probably some opportunity to make some more re-usable functions, but probably not a concern just yet. Dont want to optimize it too much this early on

Reply:
ok that's fair